bcipy/signal/model/pca_rda_kde/pca_rda_kde.py

```diff
@@ -0,0 +1,113 @@
+import pickle
+from pathlib import Path
+from typing import List
+
+import numpy as np
+from ..base_model import ModelEvaluationReport, SignalModel
+from .classifier import RegularizedDiscriminantAnalysis
+from .cross_validation import cost_cross_validation_auc, cross_validation
+from .density_estimation import KernelDensityEstimate
+from .dimensionality_reduction import ChannelWisePrincipalComponentAnalysis
+from .pipeline import Pipeline
+from scipy.stats import iqr
+
+
+class NotFittedError(Exception):
+    pass
+
+
+class PcaRdaKdeModel(SignalModel):
+    def __init__(self, k_folds=10):
+        self.k_folds = k_folds
+        self.model = None
+        self._ready_to_predict = False
+
+    def fit(self, train_data: np.array, train_labels: np.array) -> SignalModel:
+        """
+        Train on provided data using K-fold cross validation and return self.
+
+        Parameters:
+            train_data: shape (Channels, Trials, Trial_length) preprocessed data
+            train_labels: shape (Trials,) binary labels
+
+        Returns:
+            trained likelihood model
+        """
+        pca = ChannelWisePrincipalComponentAnalysis(var_tol=1e-5, num_ch=train_data.shape[0])
+        rda = RegularizedDiscriminantAnalysis()
+        model = Pipeline()
+        model.add(pca)
+        model.add(rda)
+
+        # Find the optimal gamma + lambda values
+        arg_cv = cross_validation(train_data, train_labels, model=model, k_folds=self.k_folds)
+
+        # Get the AUC using those optimized gamma + lambda
+        model.pipeline[1].lam = arg_cv[0]
+        model.pipeline[1].gam = arg_cv[1]
+        _, sc_cv, y_cv = cost_cross_validation_auc(
+            model, 1, train_data, train_labels, arg_cv, k_folds=self.k_folds, split="uniform"
+        )
+
+        # After finding cross validation scores do one more round to learn the
+        # final RDA model
+        model.fit(train_data, train_labels)
+
+        # Insert the density estimates to the model and train using the cross validated
+        # scores to avoid over fitting. Observe that these scores are not obtained using
+        # the final model
+        bandwidth = 1.06 * min(np.std(sc_cv), iqr(sc_cv) / 1.34) * np.power(train_data.shape[0], -0.2)
```

Comment:
Hmm - it's true that the exact same statement occurs a number of times throughout the current code. I suppose it's a good idea to compute it in one place to prevent bugs, even if I can't really foresee changing it

Reply:
Sigh. While refactoring this, I found that the bandwidth was actually calculated incorrectly previously. The current code preserves the way it was calculated, but it does not actually follow the "rule of thumb" that it was trying to.

https://en.wikipedia.org/wiki/Kernel_density_estimation#A_rule-of-thumb_bandwidth_estimator

The suggested rules of thumb are as follows, where:
```python
n_channels, n_items, n_samples = x.shape
```
(Note - this is a weird ordering. In pytorch/tensorflow/most numpy algorithms the batch dimension is typically kept first, and this is precisely the source of the current bug!)

Rule 1 from wikipedia:
```python
1.06 * np.std(x) * (n_items ** -0.2)
```

Rule 2 from wikipedia:
```python
0.9 * min(np.std(x), iqr(x) / 1.34) * (n_items ** -0.2)
```

Current logic (https://github.com/CAMBI-tech/BciPy/blob/main/bcipy/signal/model/mach_learning/train_model.py#L60):
```python
1.06 * min(np.std(x), iqr(x) / 1.34) * (n_channels ** -0.2)
```

The problems are:
1. It mixes the two rules of thumb (one with a multiplier of 1.06 and the other with 0.9). This is probably less important and wikipedia might have it wrong.
2. More important, it uses `n_channels` instead of `n_items`


As a takeaway, good suggestion, because it seems to have caught a bug, and DRY for the win once again.

Reply:
After discussion, we decided to maintain the current behavior and investigate this further. https://www.pivotaltracker.com/story/show/178045750